uast/diff/diff.go

```diff
@@ -0,0 +1,283 @@
+package diff
+
+import (
+	// TODO: integrate with https://github.com/src-d/lapjv if perf is unacceptable
+	"fmt"
+	"github.com/heetch/lapjv"
+	"gopkg.in/bblfsh/sdk.v2/uast/nodes"
+)
+
+type decisionType interface {
+	isDecisionType()
+}
+
+type basicDecisionType struct{}
+
+func (_ basicDecisionType) isDecisionType() {}
+
+type decision struct {
+	cost     int
+	decision decisionType
+}
+
+// match decision types together with their params
+type sameDecision struct{ basicDecisionType }
+type replaceDecision struct{ basicDecisionType }
+type matchDecision struct{ basicDecisionType }
+type permuteDecision struct {
+	basicDecisionType
+	permutation []int
+}
+
+//convinience method for choosing the cheapest decision
+func (self *decision) minEq(candidate *decision) {
+	if _, ok := candidate.decision.(replaceDecision); ok || self.cost > candidate.cost {
+		self.cost, self.decision = candidate.cost, candidate.decision
+	}
+}
+
+//type for cache
+type keyType struct{ k1, k2 ID }
+
+//cache for diff computation
+type cacheStorage struct {
+	decisions map[keyType]decision
+	sizes     map[ID]int
+	changes   Changelist
+}
+
+func emptyCacheStorage() cacheStorage {
+	return cacheStorage{
+		make(map[keyType]decision),
+		make(map[ID]int),
+		make([]Change, 0),
+	}
+}
+
+// caches (for perf) tree size (node count)
+func (ds *cacheStorage) nodeSize(n nodes.Node) int {
+	label := nodes.UniqueKey(n)
+	if cnt, ok := ds.sizes[label]; ok {
+		return cnt
+	}
+	ret := nodes.Count(n, nodes.KindsNotNil)
+	ds.sizes[label] = ret
+	return ret
+}
+
+// find the cheapest way to naively match src and dst and return the action with its combined cost
+func (ds *cacheStorage) decideAction(src, dst nodes.Node) decision {
+	label := keyType{nodes.UniqueKey(src), nodes.UniqueKey(dst)}
+
+	if val, ok := ds.decisions[label]; ok {
+		return val
+	}
+
+	// one can always just create the dst ignoring the src
+	cost := ds.nodeSize(dst) + 1
+	bestDecision := decision{cost, replaceDecision{}}
+
+	if nodes.KindOf(src) != nodes.KindOf(dst) {
+		ds.decisions[label] = bestDecision
+		return bestDecision
+	}
+
+	switch src.(type) {
```

Comment:
If written as `switch src := src.(type) {`, the `src` variable will have the right type inside each `case`, so you won't need to cast it again.

Reply:
I wanted to do a switch with both src and dst but the go wouldn't let me. Both variables have the same type, I've checked that.

Reply:
Comment is still valid - you can write `src := src.(type)` and drop type assertions on `src` (but not `dst`).

Reply:
done.